src/main/java/org/broadinstitute/hellbender/transformers/DUSTReadTransformer.java

```diff
@@ -0,0 +1,230 @@
+package org.broadinstitute.hellbender.transformers;
+
+import htsjdk.samtools.Cigar;
+import htsjdk.samtools.util.Tuple;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.cmdline.Argument;
+import org.broadinstitute.hellbender.tools.ClipReads;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import htsjdk.samtools.SAMUtils;
+
+import java.util.*;
+
+//Masks low-complexity sequences with 'N' and base qualities with '#'
+
+public class DUSTReadTransformer implements ReadTransformer {
+    public static final long serialVersionUID = 1L;
+    private final Logger logger = LogManager.getLogger(DUSTReadTransformer.class);
+
+    @Argument(fullName="Mask PHRED score", shortName="dustPhred", optional=true)
+    public short MASK_PHRED = 15;
+
+    @Argument(fullName="Window size", shortName="dustW", optional=true)
+    public int W = 64;
+
+    @Argument(fullName="Score threshold", shortName="dustT", optional=true)
+    public float T = 20;
+
+    private static final int NUM_TRIPLETS = 64;
+
+    private class PStruct {
+        public int start = 0;
+        public int finish = 0;
+        public float score = 0;
+    }
+
+    @Override
+    public GATKRead apply(final GATKRead read) {
+
+        byte[] q = read.getBases();
+
+        //Find convert to 2-bit and assign random bases to N's
+        //TODO: reject reads with too many N's
+        Random rng = new Random();
+        rng.setSeed(0);
+        for (int i = 0; i < q.length; i++) {
+            if (q[i] == 'N') {
+                q[i] = (byte)rng.nextInt(4);
+            } else {
+                if (q[i] == 'A' || q[i] == 'a') {
+                    q[i] = 0x0;
+                } else if (q[i] == 'T' || q[i] == 't') {
+                    q[i] = 0x1;
+                } else if (q[i] == 'C' || q[i] == 'c') {
+                    q[i] = 0x2;
+                } else if (q[i] == 'G' || q[i] == 'g') {
+                    q[i] = 0x3;
+                } else {
+                    logger.warn("Invalid base " + (char) q[i] + " in read " + read.getName() + ", assigning to random base.");
+                    q[i] = (byte)rng.nextInt(4);
+                }
+            }
+        }
+
+        //Start main algorithm
+        List<Tuple<Integer,Integer>> res = new ArrayList<>();
+        LinkedList<PStruct> P = new LinkedList<>();
```

Comment:
It's not ideal, but partly it's just the way Java works.  Lists passed as arguments could be modifiable or unmodifiable, and the subroutine might modify them or not, and there's nothing that signals which Lists are unmodifiable nor any way to mark which passed objects will be left unmodified.  I'd just document it and not worry about it.